File: frontend/src/features/visits/recurrence.ts
/** Mirror of backend apps.visits.recurrence — keep in sync. */

export type VisitRecurrence = "once" | "weekly" | "biweekly" | "monthly";

export const RECURRENCE_OPTIONS: { value: VisitRecurrence; label: string; hint: string }[] = [
  { value: "once", label: "One-time", hint: "Single visit on the selected date" },
  { value: "weekly", label: "Once a week", hint: "Same weekday every week" },
  { value: "biweekly", label: "Every other week", hint: "Alternate weeks from the start date" },
  { value: "monthly", label: "Same day each month", hint: "e.g. 3rd of every month" },
];

const MAX_OCCURRENCES = 52;

function parseISODate(iso: string): Date {
  const [y, m, d] = iso.split("-").map(Number);
  return new Date(y, m - 1, d);
}

function toISODate(d: Date): string {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

function addMonths(d: Date, months: number): Date {
  const year = d.getFullYear();
  const month = d.getMonth() + months;
  const day = d.getDate();
  const target = new Date(year, month, 1);
  const lastDay = new Date(target.getFullYear(), target.getMonth() + 1, 0).getDate();
  target.setDate(Math.min(day, lastDay));
  return target;
}

export function expandRecurrenceDates(
  startISO: string,
  recurrence: VisitRecurrence,
  untilISO?: string,
): string[] {
  if (!startISO) return [];
  if (recurrence === "once" || !untilISO) return [startISO];

  const start = parseISODate(startISO);
  const until = parseISODate(untilISO);
  if (until < start) return [startISO];

  const dates: string[] = [startISO];

  if (recurrence === "weekly" || recurrence === "biweekly") {
    const stepDays = recurrence === "weekly" ? 7 : 14;
    const current = new Date(start);
    while (dates.length < MAX_OCCURRENCES) {
      current.setDate(current.getDate() + stepDays);
      if (current > until) break;
      dates.push(toISODate(current));
    }
    return dates;
  }

  let offset = 1;
  while (dates.length < MAX_OCCURRENCES) {
    const current = addMonths(start, offset);
    if (current > until) break;
    dates.push(toISODate(current));
    offset += 1;
  }
  return dates;
}

export function ordinalDay(day: number): string {
  const mod100 = day % 100;
  if (mod100 >= 11 && mod100 <= 13) return `${day}th`;
  switch (day % 10) {
    case 1:
      return `${day}st`;
    case 2:
      return `${day}nd`;
    case 3:
      return `${day}rd`;
    default:
      return `${day}th`;
  }
}

export function formatRecurrenceSummary(
  startISO: string,
  recurrence: VisitRecurrence,
  untilISO?: string,
): string {
  if (!startISO) return "";
  const dates = expandRecurrenceDates(startISO, recurrence, untilISO);
  const day = parseISODate(startISO).getDate();
  const weekday = parseISODate(startISO).toLocaleDateString(undefined, { weekday: "long" });

  if (recurrence === "once") return `1 visit on ${startISO}`;
  if (!untilISO) return "Choose an end date to preview visits";

  const label =
    recurrence === "weekly"
      ? `Every ${weekday}`
      : recurrence === "biweekly"
        ? `Every other ${weekday}`
        : `${ordinalDay(day)} of each month`;

  return `${dates.length} visit${dates.length === 1 ? "" : "s"} · ${label} through ${untilISO}`;
}

File: backend/apps/visits/recurrence.py
"""Expand a visit start date into a recurrence series."""

from __future__ import annotations

from calendar import monthrange
from datetime import date, timedelta

MAX_OCCURRENCES = 52

RECURRENCE_ONCE = "once"
RECURRENCE_WEEKLY = "weekly"
RECURRENCE_BIWEEKLY = "biweekly"
RECURRENCE_MONTHLY = "monthly"

RECURRENCE_CHOICES = (
    (RECURRENCE_ONCE, "One-time"),
    (RECURRENCE_WEEKLY, "Once a week"),
    (RECURRENCE_BIWEEKLY, "Every other week"),
    (RECURRENCE_MONTHLY, "Same day each month"),
)


def _add_months(d: date, months: int) -> date:
    """Advance by N months, clamping day to the last day of the target month."""
    month_index = d.month - 1 + months
    year = d.year + month_index // 12
    month = month_index % 12 + 1
    day = min(d.day, monthrange(year, month)[1])
    return date(year, month, day)


def expand_recurrence(
    start: date,
    recurrence: str,
    until: date | None = None,
    *,
    max_occurrences: int = MAX_OCCURRENCES,
) -> list[date]:
    """
    Return scheduled dates including `start`.

    - once: [start]
    - weekly: start, then every 7 days through until
    - biweekly: start, then every 14 days through until
    - monthly: start day-of-month each month through until (clamped for short months)
    """
    if recurrence == RECURRENCE_ONCE or until is None:
        return [start]

    if until < start:
        raise ValueError("until_date must be on or after scheduled_date")

    dates: list[date] = [start]
    if recurrence == RECURRENCE_WEEKLY:
        step = timedelta(days=7)
        current = start + step
        while current <= until and len(dates) < max_occurrences:
            dates.append(current)
            current += step
    elif recurrence == RECURRENCE_BIWEEKLY:
        step = timedelta(days=14)
        current = start + step
        while current <= until and len(dates) < max_occurrences:
            dates.append(current)
            current += step
    elif recurrence == RECURRENCE_MONTHLY:
        month_offset = 1
        while len(dates) < max_occurrences:
            current = _add_months(start, month_offset)
            if current > until:
                break
            dates.append(current)
            month_offset += 1
    else:
        raise ValueError(f"Unknown recurrence: {recurrence}")

    return dates
